### ISSUE TITLE
Heap out of bounds read in bootp_print()

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#557]
**Heap out of bounds read in bootp_print()**

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r oobr_bootp_print.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/20091209/cfd01abc-a591-11e6-8591-d4e7ef382bed.png)

ASAN output:

```
==8091==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60400000dfbc at pc 0x00000053d61d bp 0x7fffb28645b0 sp 0x7fffb28645a8
READ of size 1 at 0x60400000dfbc thread T0
    #0 0x53d61c in bootp_print XYZ/tcpdump/./print-bootp.c:294:31
    #1 0x691493 in udp_print XYZ/tcpdump/./print-udp.c:551:4
    #2 0x58f360 in ip_print_demux XYZ/tcpdump/./print-ip.c:383:3
    #3 0x592006 in ip_print XYZ/tcpdump/./print-ip.c:648:3
    #4 0x566f7c in ethertype_print XYZ/tcpdump/./print-ether.c:325:10
    #5 0x566418 in ether_print XYZ/tcpdump/./print-ether.c:229:7
    #6 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #7 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #8 0x7adaf7 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #9 0x6b6f4c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #10 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #11 0x7f1a5e44b82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #12 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

0x60400000dfbc is located 0 bytes to the right of 44-byte region [0x60400000df90,0x60400000dfbc)
allocated by thread T0 here:
    #0 0x4b880c in malloc /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64:3
    #1 0x7af103 in pcap_check_header XYZ/libpcap/./sf-pcap.c:401:14
    #2 0x7acf22 in pcap_fopen_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:400:7
    #3 0x7accaf in pcap_open_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:307:6
    #4 0x7f1a5e44b82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow XYZ/tcpdump/./print-bootp.c:294:31 in bootp_print
Shadow bytes around the buggy address:
  0x0c087fff9ba0: fa fa fa fa fa fa fa fa fa fa 00 00 00 00 00 fa
  0x0c087fff9bb0: fa fa 00 00 00 00 05 fa fa fa 00 00 00 00 05 fa
  0x0c087fff9bc0: fa fa 00 00 00 00 06 fa fa fa 00 00 00 00 06 fa
  0x0c087fff9bd0: fa fa 00 00 00 00 00 fa fa fa 00 00 00 00 00 fa
  0x0c087fff9be0: fa fa fd fd fd fd fd fd fa fa fd fd fd fd fd fd
=>0x0c087fff9bf0: fa fa 00 00 00 00 00[04]fa fa 00 00 00 00 00 01
  0x0c087fff9c00: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c087fff9c40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==8091==ABORTING


```

**Comment by fxlb** [comment]
Thanks for the report.
This issue is now fixed in the security branch (Will be merged in a few weeks).

**Comment by infrastation** [comment]
You are the original reporter of this vulnerability. Commit 496be87 fixes this and the fix is available in the release 4.9.0. Thank you very much for the report, please send any future security reports to security@tcpdump.org.
